### ISSUE TITLE
Flatpak applications using the SecretService API show up as xdg-dbus-proxy

### ISSUE THREAD

**Issue by ntninja** [keepassxreboot/keepassxc#8946]
## Summary
After #3837 all binaries accessing the KeePassXC store using SecretService now need to be confirmed by a permission prompt. Unfortunately all sandboxed Flatpak applications are displayed as accessing KeePassXC as **/usr/bin/xdg-dbus-proxy (PID: <pid>)** with any sandboxed application being able to access the key material for all other sandboxed applications. (The issue is probably similar for non-sandboxed applications developed in scripting languages, but unlike those it’s actually possible to reliable determine with sandboxed app the call originates from so it can actually be addressed in a proper way.)

![Screenshot_20230101_231144](https://user-images.githubusercontent.com/246386/210186054-5285354f-6099-41cf-bff1-38c081229c51.png)

## Context
Congratulations on adding #3837 to the codebase more than a year ago, Debian just got it into unstable!

To get the Flatpak App-ID for PID, read the file */proc/<pid>/root/.flatpak-info* as INI (GKeyFile to be exact) and look for the “name” key in the “Application” or “Runtime” section. If the file does not exist, it is not a sandboxed application. The xdg-desktop-portal source code has the exact steps to perform here: https://github.com/flatpak/xdg-desktop-portal/blob/e5153d402d981e3b1f49babaaad0da49e3d4eaa8/src/xdp-utils.c#L575-L681 (C using GLib).

Looking up a Name an Icon for an App-ID is a surprisingly hard endeavour however. Looking up its source, xdg-desktop-portal constructs a .desktop-filename from the App-ID and looks that up using `g_desktop_app_info_new` (from the GLib Gio-Unix library) to the system launcher/menu text associated with the app… pretty easy to do for a GLib-based app, but unfortunately disregards the fact that Flatpak apps are not actually required to include a .desktop-file by that name in the first place. A more robust approach would be to either:

  1. Use libappstreamqt to look up the software repository metadata for the installed application which all Flatpak applications *have* to include (there is an equivalent Qt API, but its documentation does not appear to be published online):
     1. Call [`as_pool_load`](https://www.freedesktop.org/software/appstream/docs/api/appstream-AsPool.html#as-pool-load)
     2. Call [`as_pool_get_components_by_id`](https://www.freedesktop.org/software/appstream/docs/api/appstream-AsPool.html#as-pool-get-components-by-id) on the result of the previous call using the App-ID as a parameter
     3. Call [`as_component_get_pkgname`](https://www.freedesktop.org/software/appstream/docs/api/appstream-AsComponent.html#as-component-get-pkgname) on each of the results of the previous call until a component whose pkgname matches the following is found:  
         “flatpak:” + (“app” if App-ID lookup found the ID in an *Application* section, “runtime” otherwise) + “/“ + the App-ID + “/“ + the *arch* field from the *Instance* section of the manifest + “/“ + the *branch* field from the *Instance* section of the manifest
     4. Finally, call [`as_component_get_name`](https://www.freedesktop.org/software/appstream/docs/api/appstream-AsComponent.html#as-component-get-name) and [`as_component_get_icons`](https://www.freedesktop.org/software/appstream/docs/api/appstream-AsComponent.html#as-component-get-icons) to get the wanted metadata
 2. Use `libflatpak`, but that requires figuring out which installation directory was used for the application and *that* information can only be guessed from the information available in the file that was used for the App-ID lookup. If one *does* guess the path based on what is available in that file one can (just C API in this case):
    1. Call [`flatpak_installation_new_for_path`](https://docs.flatpak.org/en/latest/libflatpak-api-reference.html#flatpak-installation-new-for-path)
    2. Call [`flatpak_installation_get_installed_ref`](https://docs.flatpak.org/en/latest/libflatpak-api-reference.html#flatpak-installation-get-installed-ref) on the result with lookup parameters matching those of 1.3.
    3. Call [`flatpak_installed_ref_get_appdata_name`](https://docs.flatpak.org/en/latest/libflatpak-api-reference.html#flatpak-installed-ref-get-appdata-name) to get the display name – I just realized this does not actually appear to have any way to look up the icon.

Considering how hard it is to reliably convert an App-ID to an application name, it might be best to skip this part entirely and just display (and store!) the App-ID for sandboxed applications accessing KeepassXC using the SecretService API.

**Comment by scabala** [comment]
I think reading `/proc/<PID>/root/.flatpak-info` would not work for KeePassXC packaged as Flatpak itself. Maybe there's a some D-Bus API in Flatpak that would allow to lookup and app given the PID?

**Comment by ntninja** [comment]
You are correct on that, Flatpak apps run inside a PID namespace, hence do not 
have access to process information outside their own namespace.

The only “D-Bus API” available would be to use
“org.freesktop.Flatpak/org/freedesktop/Flatpak/Development/HostCommand”
(aka full host access) to spawn `cat "/proc/<PID>/root/.flatpak-info"` outside 
the sandbox and read the result. At that point you no longer have any sandbox, 
since that API (that also powers `flatpak-spawn --host`) also an application 
to perform any action that the current user is allowed to perform.

It would work to get access to `.flatpak-info` inside the (no-longer) sandbox 
though.

Generally I’ve been mostly considering unsandboxed KeePassXC for this feature 
though – having it be sandboxed while attempting these kinds of deep system 
integration things is always going to have issues.


Am Donnerstag, 30. Oktober 2025, 11:53:58 Mitteleuropäische Normalzeit schrieb 
scabala:
> scabala left a comment (keepassxreboot/keepassxc#8946)
> 
> I think reading `/proc/<PID>/root/.flatpak-info` would not work for
> KeePassXC packaged as Flatpak itself. Maybe there's a some D-Bus API in
> Flatpak that would allow to lookup and app given the PID?

**Comment by scabala** [comment]
I agree that `flatpak-spawn --host` route is a big no-no. However, if such API would exist, it would have a benefit for working for both sandboxed and natively installed KeePassXC. I'll try to dig for it in `xdg-desktop-portal` or Flatpak in general.

**Comment by scabala** [comment]
After some surface search, it essentially boils down to [this](https://gitlab.freedesktop.org/dbus/dbus/-/issues/171) issue on D-Bus. 

If I understand correctly, also `/proc` approach would not be considered secure, or secure _enough_, for portal developers to take into account. I was actually about asking to implement a new portal like CallingAppDiscovery - something related but not exactly the same was described [here](https://github.com/flatpak/xdg-desktop-portal/issues/283) - but I doubt there will be any interest on implementing it with only viable option at the moment with is reading `/proc`.

Sadly, it looks like this very nice feature could not be implemented any time soon.

**Comment by ntninja** [comment]
No, the `/proc` approach is kinda wonkey, but it’s not insecure in the face of 
namespaces and `PR_SET_NO_NEW_PRIVS`. With this combination a queried 
application may re-exec, but will be unable to leave the mount-namespace or 
rewrite the `.flatpak-info` file, hence there is no way to fake a different 
sandboxing context between issuing the D-Bus call and the remote application 
reading that file.

It’s not a pretty solution though, that’s for sure… 😅


Am Donnerstag, 30. Oktober 2025, 22:10:31 Mitteleuropäische Normalzeit schrieb 
scabala:
> scabala left a comment (keepassxreboot/keepassxc#8946)
> 
> After some surface search, it essentially boils down to
> [this](https://gitlab.freedesktop.org/dbus/dbus/-/issues/171) issue on
> D-Bus.
> 
> If I understand correctly, also `/proc` approach would not be considered
> secure, or secure _enough_, for portal developers to take into account. I
> was actually about asking to implement a new portal like
> CallingAppDiscovery - something related but not exactly the same was
> described [here](https://github.com/flatpak/xdg-desktop-portal/issues/283)
> - but I doubt there will be any interest on implementing it with only
> viable option at the moment with is reading `/proc`.
> 
> Sadly, it looks like this very nice feature could not be implemented any
> time soon.

**Comment by scabala** [comment]
It looks like there's some hope as Flatpak itself is interested in application identification - if implemented, maybe KeePassXC could use it?

Ref: https://blog.sebastianwick.net/posts/so-peerpidfd-gets-more-useful/

**Comment by ntninja** [comment]
That might make it easier to figure out the application ID, but (as before) you still need to figure the application name/icon from that if you want it pretty. For the KeePassXC-in-Flatpak problem, you’s also need `--filesystem=/sys/fs/cgroup:ro` (if that’s a thing), to access the CGroup XAttrs from within the sandbox. Then you also need also need `--filesystem=/usr/share/metainfo:ro --filesystem=/usr/share/icons:ro --filesystem=/var/lib/flatpak/exports/share/metainfo:ro --filesystem=/var/lib/flatpak/exports/share/icons:ro  --filesystem=xdg-data/flatpak/exports/share/metainfo:ro --filesystem=xdg-data/flatpak/exports/share/icons:ro` (and no clue for snappy apps) to get access to host application info.

It’s good for avoiding `--talk-name=org.freedesktop.Flatpak` (full host access) though.
